File: src/cc/cc4py/test_2.cc
#include "cc4py_2.h"
#include "test_common.h"
#include "pybind11/stl.h"

#include <cassert>
#include <chrono>
#include <cmath>
#include <exception>
#include <iostream>
#include <memory>
#include <mutex>
#include <random>
#include <string>
#include <thread>
#include <tuple>
#include <vector>


//std::mutex cout_lock;


void do_one_thread(
    cc4py::Driver * driver,
    Datapoint const * data_in,
    long * data_out,
    int idx_from,
    int idx_to)
{
    for (int idx=idx_from; idx < idx_to; idx++) {
        long key = 0;
        while (1) {
            auto receipt = driver->submit(
                data_in[idx].float_features,
                data_in[idx].str_features,
                data_in[idx].int_feature
            );
            if (receipt.flag == 0) {
                key = receipt.key;
                break;
            } else {
                wait(1);
            }
        }
        if (key == 0) {
            data_out[idx] = -1;
            continue;
        }
        wait(1);

        while (1) {
            auto result = driver->retrieve(key);
            if (result.flag == 0) {
                data_out[idx] = result.value;
                break;
            } else {
                wait(1);
            }
        }
    }
}


int main(int const argc, char const * const * const argv)
{
    do_threads<cc4py::Driver>(do_on_thread);
    return 0;
}

File: tests/pycc/test_2.py
import time

from pycc.cpp_2 import Driver
from .common import do_threads


def do_one_thread(driver, data, results, idx_from, idx_to):
    for idx in range(idx_from, idx_to):
        while True:
            future, msg = driver.submit(**data[idx])
            if future is None:
                if msg == '':
                    time.sleep(0.001)
                else:
                    raise Exception(msg)
            else:
                break

        time.sleep(0.001)

        while True:
            if future.ready():
                val, msg = future.get(0)
                if val is None:
                    raise Exception(msg)
                results[idx] = val
                break
            else:
                time.sleep(0.001)


def test():
    do_threads(Driver, do_one_thread)

